File: apps/api/app/schemas/store.py
"""Schemas: Loja (produtos do dono) e Centro de Treinamento (CT)."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.user import UserStateOut

Currency = Literal["silver", "gold"]
CtKind = Literal["beach", "indoor"]


class ProductOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None = None
    item_type: str
    quantity: int
    price_silver: int
    price_gold: int
    image_url: str | None = None
    active: bool
    # Enriquecidos a partir do catálogo (engine/ct.py).
    item_label: str = ""
    item_emoji: str = ""


class ProductCreate(BaseModel):
    name: str = Field(min_length=1, max_length=60)
    description: str | None = Field(default=None, max_length=240)
    item_type: str
    quantity: int = Field(default=1, ge=1, le=999)
    price_silver: int = Field(default=0, ge=0, le=100_000_000)
    price_gold: int = Field(default=0, ge=0, le=1_000_000)
    image_url: str | None = Field(default=None, max_length=400)
    active: bool = True


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=60)
    description: str | None = Field(default=None, max_length=240)
    item_type: str | None = None
    quantity: int | None = Field(default=None, ge=1, le=999)
    price_silver: int | None = Field(default=None, ge=0, le=100_000_000)
    price_gold: int | None = Field(default=None, ge=0, le=1_000_000)
    image_url: str | None = Field(default=None, max_length=400)
    active: bool | None = None


class BuyRequest(BaseModel):
    product_id: uuid.UUID
    currency: Currency = "silver"


class InventoryItemOut(BaseModel):
    item_type: str
    label: str
    emoji: str
    quantity: int


class BuyResult(BaseModel):
    state: UserStateOut
    inventory: list[InventoryItemOut]
    granted_item: str
    granted_qty: int


# --- Centro de Treinamento -------------------------------------------------
class RequirementOut(BaseModel):
    item_type: str
    label: str
    emoji: str
    required: int
    owned: int
    ok: bool


class CenterOut(BaseModel):
    kind: CtKind
    label: str
    built: bool
    built_at: datetime | None = None
    requirements: list[RequirementOut]
    can_build: bool


class BuildResult(BaseModel):
    center: CenterOut
    centers: list[CenterOut]
    inventory: list[InventoryItemOut]
    message: str

File: packages/shared/src/store.ts
// Ecossistema Loja -> Centro de Treinamento (CT).
import type { UserState } from "./user";

export type CtItemType =
  | "terreno"
  | "areia"
  | "poste"
  | "refletor"
  | "mastro"
  | "rede"
  | "bola"
  | "ginasio"
  | "piso";

export type CtKind = "beach" | "indoor";
export type Currency = "silver" | "gold";

/** Catálogo de itens (rótulo + emoji) — espelha engine/ct.py do backend. */
export const CT_ITEMS: { type: CtItemType; label: string; emoji: string }[] = [
  { type: "terreno", label: "Terreno", emoji: "🟫" },
  { type: "areia", label: "Areia", emoji: "🏖️" },
  { type: "poste", label: "Poste", emoji: "🗼" },
  { type: "refletor", label: "Refletor", emoji: "💡" },
  { type: "mastro", label: "Mastro", emoji: "⛵" },
  { type: "rede", label: "Rede", emoji: "🥅" },
  { type: "bola", label: "Bola", emoji: "🏐" },
  { type: "ginasio", label: "Ginásio", emoji: "🏟️" },
  { type: "piso", label: "Piso", emoji: "🔲" },
];

export const CT_ITEM_LABEL: Record<CtItemType, string> = CT_ITEMS.reduce(
  (acc, i) => ({ ...acc, [i.type]: i.label }),
  {} as Record<CtItemType, string>,
);

export const CT_ITEM_EMOJI: Record<CtItemType, string> = CT_ITEMS.reduce(
  (acc, i) => ({ ...acc, [i.type]: i.emoji }),
  {} as Record<CtItemType, string>,
);

export const CT_KIND_LABEL: Record<CtKind, string> = {
  beach: "CT de Praia",
  indoor: "CT de Quadra",
};

export interface StoreProduct {
  id: string;
  name: string;
  description: string | null;
  item_type: CtItemType;
  quantity: number;
  price_silver: number;
  price_gold: number;
  image_url: string | null;
  active: boolean;
  item_label: string;
  item_emoji: string;
}

export interface InventoryItem {
  item_type: CtItemType;
  label: string;
  emoji: string;
  quantity: number;
}

export interface Requirement {
  item_type: CtItemType;
  label: string;
  emoji: string;
  required: number;
  owned: number;
  ok: boolean;
}

export interface TrainingCenter {
  kind: CtKind;
  label: string;
  built: boolean;
  built_at: string | null;
  requirements: Requirement[];
  can_build: boolean;
}

export interface BuyResult {
  state: UserState;
  inventory: InventoryItem[];
  granted_item: CtItemType;
  granted_qty: number;
}

export interface BuildResult {
  center: TrainingCenter;
  centers: TrainingCenter[];
  inventory: InventoryItem[];
  message: string;
}

// --- Câmbio de moedas (1 ouro = 10 prata) ---
/** to_silver: gasta ouro → recebe prata. to_gold: gasta prata → recebe ouro. */
export type ExchangeDirection = "to_silver" | "to_gold";

export const SILVER_PER_GOLD = 10;

export interface ExchangeResult {
  state: UserState;
  silver_delta: number;
  gold_delta: number;
}
